### PR TITLE
Transactional outbox

### PR DESCRIPTION
**Pull request by gtoselli**

### CODE REVIEW

File: packages/ddd-toolkit/src/repo/mongo-aggregate-repo.ts [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
 import { ISerializer } from './serializer.interface';
 import { merge } from 'lodash';
 import { DuplicatedIdError, OptimisticLockError, RepoHookError } from '../errors';
 import { ILogger } from '../logger';
 import { IInit } from '../init.interface';
+import { IEvent } from '../event-bus/event-bus.interface';
+import { IOutbox } from '../outbox/outbox.interface';
 
 export interface IAggregateRepo<A> {
     // TODO add id as a generic type
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/repo/mongo-aggregate-repo.ts:12]: remove this line

File: packages/ddd-toolkit/src/repo/mongo-aggregate-repo.ts [hunk 2]
```diff
@@ -1,17 +1,23 @@
...
     getById: (id: string) => Promise<WithVersion<A> | null>;
     save: (aggregate: A) => Promise<void>;
 }
 
+export interface IAggregateRepoWithOutbox<A> extends IAggregateRepo<A> {
+    saveAndPublish: (aggregate: A, eventsToBePublished?: IEvent<unknown>[]) => Promise<void>;
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/repo/mongo-aggregate-repo.ts:18]: rename in eventsToPublish

File: packages/ddd-toolkit/src/repo/mongo-aggregate-repo.ts [hunk 3]
```diff
@@ -21,15 +27,19 @@ type WithOptionalVersion<T> = T & { __version?: number };
 // TODO probably we should create a dedicated interface whit like DocumentWithIdAndTimestamps
 const MONGODB_UNIQUE_INDEX_CONSTRAINT_ERROR = 11000;
 
-export class MongoAggregateRepo<A, AM extends DocumentWithId> implements IAggregateRepo<A>, IInit {
+export class MongoAggregateRepo<A, AM extends DocumentWithId>
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/repo/mongo-aggregate-repo.ts:30]: create another class

File: packages/ddd-toolkit/src/outbox/mongo-outbox.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import { difference, intersection } from 'lodash';
+import { IOutbox } from './outbox.interface';
+
+type OutboxEventModel = {
+    event: IEvent<unknown>;
+    scheduledAt: Date;
+    status: 'scheduled' | 'processing' | 'published';
+    publishedAt?: Date;
+    contextName?: string;
+    locked?: boolean;
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/outbox/mongo-outbox.ts:14]: to remove

File: packages/ddd-toolkit/src/outbox/mongo-outbox.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+};
+
+export class MongoOutbox implements IOutbox {
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/outbox/mongo-outbox.ts:17]: add init interface

File: packages/ddd-toolkit/src/outbox/mongo-outbox.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                status: 'scheduled',
+                contextName: this.contextName,
+            })),
+            { session: clientSession },
+        );
+        this.logger.debug(`Scheduled events ${Object.values(insertedIds).join(', ')}`);
+        return Object.values(insertedIds).map((id) => id.toString());
+    }
+
+    public async publishEvents(eventIds: string[]): Promise<void> {
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/outbox/mongo-outbox.ts:57]: better name?

**Comment by lucagiove** [comment]
Repo with Outbox has a dedicated class that extends base MongoAggregateRepo class

File: packages/ddd-toolkit/src/repo/mongo-aggregate-repo-with-outbox.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        const aggregateVersion = aggregate.__version || 0;
+
+        const session = this.mongoClient.startSession();
+
+        const scheduledEventIds: string[] = [];
+        try {
+            await session.withTransaction(async () => {
+                await this.upsertWriteModel(aggregateModel, aggregateVersion, session);
+                await this.handleRepoHooks(aggregateModel, session);
+                await this.handleOutbox(eventsToBePublished, session);
```
Comment (gtoselli) [inline packages/ddd-toolkit/src/repo/mongo-aggregate-repo-with-outbox.ts:40]: @lucagiove should be _scheduledEventIds = await this.handleOutbox(eventsToBePublished, session);_

**Review by lucagiove — APPROVED** [review]

**Pull request merged**